File: scripts/check_secrets.py
from __future__ import annotations

import os
import re
import sys
from pathlib import Path

ROOT = Path.cwd()
SKIP_DIRS = {".git", ".venv", "node_modules", "var"}
SKIP_FILES = {"uv.lock"}
PATTERNS = [
    ("OpenAI-style key", re.compile(r"\bsk-[A-Za-z0-9_-]{12,}\b")),
    ("GitHub token", re.compile(r"\bgh[pousr]_[A-Za-z0-9_]{12,}\b")),
    ("Google API key", re.compile(r"\bAIza[0-9A-Za-z\-_]{20,}\b")),
    ("Bearer token", re.compile(r"Bearer\s+[A-Za-z0-9._-]{20,}")),
]


def iter_files(root: Path):
    for dir_path, dir_names, file_names in os.walk(root, topdown=True):
        dir_names[:] = [name for name in dir_names if name not in SKIP_DIRS]

        for file_name in file_names:
            if file_name in SKIP_FILES:
                continue

            yield Path(dir_path) / file_name


def line_number_at(content: str, index: int) -> int:
    return content.count("\n", 0, index) + 1


def main() -> int:
    violations: list[tuple[str, int, str, str]] = []

    for file_path in iter_files(ROOT):
        try:
            content = file_path.read_text(encoding="utf-8")
        except UnicodeDecodeError:
            continue

        for name, pattern in PATTERNS:
            for match in pattern.finditer(content):
                line = line_number_at(content, match.start())
                masked = match.group(0)[:16] + "..."
                violations.append((str(file_path.relative_to(ROOT)), line, name, masked))

    if not violations:
        print("Secret scan passed.")
        return 0

    print("Potential secrets detected:", file=sys.stderr)
    for rel, line, kind, masked in violations:
        print(f"- {rel}:{line} {kind} {masked}", file=sys.stderr)
    return 1


if __name__ == "__main__":
    raise SystemExit(main())

File: legacy/node-mvp/scripts/check-secrets.js
import fs from "node:fs/promises";
import path from "node:path";
import process from "node:process";

const ROOT = process.cwd();
const SKIP_DIRS = new Set([".git", "node_modules", "var"]);
const SKIP_FILES = new Set(["package-lock.json"]);
const PATTERNS = [
  { name: "OpenAI-style key", regex: /\bsk-[A-Za-z0-9_-]{12,}\b/g },
  { name: "GitHub token", regex: /\bgh[pousr]_[A-Za-z0-9_]{12,}\b/g },
  { name: "Google API key", regex: /\bAIza[0-9A-Za-z\-_]{20,}\b/g },
  { name: "Bearer token", regex: /Bearer\s+[A-Za-z0-9._-]{20,}/g }
];

async function walk(dirPath, files = []) {
  const entries = await fs.readdir(dirPath, { withFileTypes: true });

  for (const entry of entries) {
    if (entry.isDirectory()) {
      if (SKIP_DIRS.has(entry.name)) {
        continue;
      }

      await walk(path.join(dirPath, entry.name), files);
      continue;
    }

    if (SKIP_FILES.has(entry.name)) {
      continue;
    }

    files.push(path.join(dirPath, entry.name));
  }

  return files;
}

function lineNumberAt(content, index) {
  return content.slice(0, index).split(/\r?\n/).length;
}

async function main() {
  const files = await walk(ROOT);
  const violations = [];

  for (const filePath of files) {
    const relativePath = path.relative(ROOT, filePath);
    const content = await fs.readFile(filePath, "utf8");

    for (const pattern of PATTERNS) {
      const matches = content.matchAll(pattern.regex);

      for (const match of matches) {
        const line = lineNumberAt(content, match.index ?? 0);
        violations.push({
          path: relativePath,
          line,
          kind: pattern.name,
          value: match[0].slice(0, 16) + "..."
        });
      }
    }
  }

  if (violations.length === 0) {
    console.log("Secret scan passed.");
    return;
  }

  console.error("Potential secrets detected:");
  for (const violation of violations) {
    console.error(`- ${violation.path}:${violation.line} ${violation.kind} ${violation.value}`);
  }

  process.exitCode = 1;
}

main().catch((error) => {
  console.error(error instanceof Error ? error.stack : error);
  process.exitCode = 1;
});
